accepted/2020/feature-switch.md

```diff
@@ -0,0 +1,113 @@
+
+*Note that it seems likely that some of switches may require 1:many mapping because they're targeting existing code which uses multiple properties to determine the presence of the feature.*
+
+Ultimately some of these switches may also need VS UI integration, but for now most of the feature switches should be perfectly fine with only MSBuild properties.
+
+The ability to specify these from MSBuild projects as properties also means that different templates and SDKs can set different defaults - so for example Blazor or Xamarin SDKs may choose to turn off some of these features by default.
+
+### Generate the right input for the linker in SDK
+Currently there's no integration of the substitutions feature in the linker with the `ILLink` task/targets. Introduce a standard way to  define the content of `substitutions.xml` from the SDK (MSBuild items/properties) and then implement the file generation in the `ILLink` task and pass it to the linker.
+
+*Or change linker to not use the XML format and instead take the input on command line.*
+
+
+## Open questions
+
+### Boolean only or more complex values
+Should the feature switches always represent simple on/off statements and thus be represented as boolean properties? In some cases the simple boolean may necessitate presence of multiple similar feature switches. For example the security algorithms case described above. In the extreme each such algorithm would have its own feature switch - this would obviously create confusion. We could potentially allow feature switches to have string values from a predefined set to solve this problem.
+
+### Which components own the mapping between different representations of feature switches
+There are 3 different representations to each feature switch
+* The MSBuild property - for example `InvariantGlobalization` - this is used by developers to turn the feature on/off.
+* The runtime property in `.runtimeconfig.json` - for example `System.Globalization.Invariant` - this is used at runtime to alter the behavior of the code (via the IL property below)
+* The IL property which is used in the code - for example `System.Globalization.GlobalizationMode.Invariant, System.Private.CoreLib` - this is used by the linker to fix the value of the property and propagate that to remove unused code.
+
+The mapping between MSBuild property and runtime property pretty much has to be part of the SDK (in the MSBuild somewhere) as the SDK already generates the `.runtimeconfig.json` and this needs to be part of it. The second mapping to the IL property can take several options:
+* Also int he SDK - becomes an input to the ILLink task
+* Encoded as attributes in the BCL - we could add a new attribute to the IL properties to mark which runtime options they map to - linker can create the mapping from these attributes. In this case the ILLink task would take the runtime properties as input
```

Comment:
This would mean the libraries would need to know about the MSBuild property names?

Reply:
The idea is to use the runtime property names in the attributes - for example:
```C#
internal static class GlobalizationMode
{
    [FeatureSwitch("System.Globalization.Invariant")]
    internal static bool Invariant { get; }
}
```

Reply:
In my mind, the name that is the "glue" is the string that is in the `runtimeconfig.json`. This is the "Public API" for this switch from the runtime/library perspective. This is the string that the library passes into `AppContext.TryGetSwitch`, so the library already knows about this string.

On the MSBuild side, there is a step in-between the property name and the `runtimeconfig.json`, and that is the `RuntimeHostConfigurationOption` item.  `RuntimeHostConfigurationOption` is the information that gets put into the `runtimeconfig.json`. The mapping between the MSBuild property and the `RuntimeHostConfigurationOption` also lives in MSBuild:

https://github.com/dotnet/sdk/blob/36ef8b2aa8e5d579c921704bdab69a7407936889/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L341-L354

```xml
  <!--
    Defaults @(RuntimeHostConfigurationOption) items based on MSBuild properties.
    -->

    <RuntimeHostConfigurationOption Include="System.Globalization.Invariant"
                                    Condition="'$(InvariantGlobalization)' != ''"
                                    Value="$(InvariantGlobalization)" />

    <RuntimeHostConfigurationOption Include="System.GC.Server"
                                    Condition="'$(ServerGarbageCollection)' != ''"
                                    Value="$(ServerGarbageCollection)" />
```

Reply:
@eerhardt Exactly - "runtime property" == "property in `.runtimeconfig.json`". Sorry for the confusion in terminology.